app/src/main/java/org/mozilla/fenix/historymetadata/HistoryMetadataMiddleware.kt

```diff
@@ -0,0 +1,128 @@
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        // Pre process actions
+        when (action) {
+            is TabListAction.AddTabAction -> {
+                if (action.select) {
+                    // Before we add and select a new tab we update the metadata
+                    // of the currently selected tab, if non-private.
+                    context.state.selectedNormalTab?.let {
+                        updateHistoryMetadata(it)
+                    }
+                }
+            }
+            is TabListAction.SelectTabAction -> {
+                // Before we select a new tab we update the metadata
+                // of the currently selected tab, if non-private.
+                context.state.selectedNormalTab?.let {
+                    updateHistoryMetadata(it)
+                }
+            }
+            is TabListAction.RemoveTabAction -> {
+                if (action.tabId == context.state.selectedTabId) {
+                    context.state.findNormalTab(action.tabId)?.let {
+                        updateHistoryMetadata(it)
+                    }
+                }
+            }
+            is TabListAction.RemoveTabsAction -> {
+                action.tabIds.find { it == context.state.selectedTabId }?.let {
+                    context.state.findNormalTab(it)?.let { tab ->
+                        updateHistoryMetadata(tab)
+                    }
+                }
+            }
+            is ContentAction.UpdateLoadingStateAction -> {
+                context.state.findNormalTab(action.sessionId)?.let { tab ->
+                    val selectedTab = tab.id == context.state.selectedTabId
+                    if (tab.content.loading && !action.loading) {
+                        // When a page stops loading we record its metadata
+                        createHistoryMetadata(context, tab)
+                    } else if (!tab.content.loading && action.loading && selectedTab) {
+                        // When a page starts loading (e.g. user navigated away by
+                        // clicking on a link) we update metadata
+                        updateHistoryMetadata(tab)
+                    }
+                }
+            }
+        }
+
+        next(action)
+
+        // Post process actions
+        when (action) {
+            // We're handling this after processing the action because we want the tab
+            // state to contain the updated media session state.
+            is MediaSessionAction.UpdateMediaMetadataAction -> {
+                context.state.findNormalTab(action.tabId)?.let { tab ->
+                    createHistoryMetadata(context, tab)
+                }
+            }
+        }
+    }
+
+    private fun createHistoryMetadata(context: MiddlewareContext<BrowserState, BrowserAction>, tab: TabSessionState) {
+        val key = historyMetadataService.createMetadata(tab, tab.getParent(context.store))
```

Comment:
Should this be better named `createMetadataKey`?

Reply:
@gabrielluong Ah no, so this actually stores/creates a metadata (not just the key). It records a document type observation and returns the key. The key is the set of attributes that uniquely identify the metadata of this tab. Naming is not ideal yet but `createMetadata` is the best we could come up with for now.